### PR TITLE
[CI Environment] Enable 'Publish if not exists' + Fix for 'Publish latest'

### PR DESCRIPTION
**Pull request by AlexanderSehr**
# Description

In case people onboard CARML directly into main they have to apply workarounds to get their modules to publish as the CI envrionment expects a merge to main with module changes in order to publish.

This PR mitigates this issue but not only searching for changes, but also checking of any given module (or child-module) is already part of the target environment (e.g. Bicep Registry, Template Specs, Azure DevOps Artifacts Feed).

Also, this PR fixes an issue with the 'Publish latest' in that it adds a missing passthru of the settings.yml variable to the actual function that needs it.

Required tests:
| Test Case \ Environment | Azure DevOps | GitHub |
| - | - | - |
| Target not existing Artifacts | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=2&branchName=users%2Falsehr%2F2153_PublishIfNotExists&buildid=16783) | - |
| Target not existing ACR| [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=2&branchName=users%2Falsehr%2F2153_PublishIfNotExists&buildid=16783) | [AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/runs/3639224056/attempts/1) |
| Target not existing TS | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=2&branchName=users%2Falsehr%2F2153_PublishIfNotExists&buildid=16783) | [AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/runs/3639224056/attempts/1) |
| Target existing Artifacts | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=2&branchName=users%2Falsehr%2F2153_PublishIfNotExists&buildid=16784) | - |
| Target existing ACR| [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=2&branchName=users%2Falsehr%2F2153_PublishIfNotExists&buildid=16784) | [AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/runs/3639224056/attempts/2) |
| Target existing TS | [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=2&branchName=users%2Falsehr%2F2153_PublishIfNotExists&buildid=16784) | [AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/runs/3639224056/attempts/2) |

The idea for a 'first' run in an 'empty' environment is to see an output such as

<details>
<summary>Bicep registry example</summary>

```
Publish module to private bicep registry
  VERBOSE: Invoke task with
  VERBOSE: {
    "PublishLatest": true,
    "TemplateFilePath": "/home/runner/work/ResourceModules/ResourceModules/modules/Microsoft.AnalysisServices/servers/deploy.bicep"
  }
  
  VERBOSE: Gathering modified files between current branch and main
  VERBOSE: Looking for modified files under: [Microsoft.AnalysisServices/servers]
  VERBOSE: No template file found in the modified module.
  VERBOSE: Modified modules found: [0]
  VERBOSE: No modules with changes found to publish.
  VERBOSE: Invoke Get-ModulesMissingFromPrivateBicepRegistry with
  VERBOSE: {
    "PublishLatest": true,
    "BicepRegistryRgName": "artifacts-rg",
    "BicepRegistryName": "adpsxxazacrx001",
    "TemplateFilePath": "/home/runner/work/ResourceModules/ResourceModules/modules/Microsoft.AnalysisServices/servers/deploy.bicep"
  }
  
  VERBOSE: Missing module [/home/runner/work/ResourceModules/ResourceModules/modules/Microsoft.AnalysisServices/servers/deploy.bicep] will be considered for publishing with version(s) [0.4.0, 0.4, 0, latest]
  VERBOSE: No modules missing in the target environment
 - [Microsoft.AnalysisServices/servers] [0.4.0]
 - [Microsoft.AnalysisServices/servers] [0.4]
  VERBOSE: Invoke task with
  VERBOSE: {
    "BicepRegistryRgName": "artifacts-rg",
    "ModuleVersion": "0.4",
    "BicepRegistryName": "adpsxxazacrx001",
    "BicepRegistryRgLocation": "West Europe",
    "TemplateFilePath": "/home/runner/work/ResourceModules/ResourceModules/modules/Microsoft.AnalysisServices/servers/deploy.bicep"
  }
  
  VERBOSE: Performing the operation "Publish" on target "Private bicep registry entry [bicep/modules/microsoft.analysisservices.servers] version [0.4] to registry [adpsxxazacrx001]".
  VERBOSE: Publish complete
 - [Microsoft.AnalysisServices/servers] [0]
  VERBOSE: Invoke task with
  VERBOSE: {
    "BicepRegistryRgName": "artifacts-rg",
    "ModuleVersion": "0",
    "BicepRegistryName": "adpsxxazacrx001",
    "BicepRegistryRgLocation": "West Europe",
    "TemplateFilePath": "/home/runner/work/ResourceModules/ResourceModules/modules/Microsoft.AnalysisServices/servers/deploy.bicep"
  }
  
  VERBOSE: Performing the operation "Publish" on target "Private bicep registry entry [bicep/modules/microsoft.analysisservices.servers] version [0] to registry [adpsxxazacrx001]".
  VERBOSE: Publish complete
 - [Microsoft.AnalysisServices/servers] [latest]
  VERBOSE: Invoke task with
  VERBOSE: {
    "BicepRegistryRgName": "artifacts-rg",
    "ModuleVersion": "latest",
    "BicepRegistryName": "adpsxxazacrx001",
    "BicepRegistryRgLocation": "West Europe",
    "TemplateFilePath": "/home/runner/work/ResourceModules/ResourceModules/modules/Microsoft.AnalysisServices/servers/deploy.bicep"
  }
  
  VERBOSE: Performing the operation "Publish" on target "Private bicep registry entry [bicep/modules/microsoft.analysisservices.servers] version [latest] to registry [adpsxxazacrx001]".
  VERBOSE: Publish complete
```

</details>

Then, for any subsequent run, the output should look similar to

<details>
<summary>Bicep registry example</summary>

```
Publish module to private bicep registry
  VERBOSE: Invoke task with
  VERBOSE: {
    "TemplateFilePath": "/home/runner/work/ResourceModules/ResourceModules/modules/Microsoft.AnalysisServices/servers/deploy.bicep",
    "PublishLatest": true
  }
  
  VERBOSE: Gathering modified files between current branch and main
  VERBOSE: Looking for modified files under: [Microsoft.AnalysisServices/servers]
  VERBOSE: No template file found in the modified module.
  VERBOSE: Modified modules found: [0]
  VERBOSE: No modules with changes found to publish.
  VERBOSE: Invoke Get-ModulesMissingFromPrivateBicepRegistry with
  VERBOSE: {
    "PublishLatest": true,
    "BicepRegistryRgName": "artifacts-rg",
    "TemplateFilePath": "/home/runner/work/ResourceModules/ResourceModules/modules/Microsoft.AnalysisServices/servers/deploy.bicep",
    "BicepRegistryName": "adpsxxazacrx001"
  }
  
  VERBOSE: No modules missing in the target environment
```

</details>

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: .github/workflows/ms.aad.domainservices.yml [hunk 1]
```diff
@@ -142,3 +142,4 @@ jobs:
           bicepRegistryRGName: '${{ env.bicepRegistryRGName }}'
           bicepRegistryRgLocation: '${{ env.bicepRegistryRgLocation }}'
           bicepRegistryDoPublish: '${{ env.bicepRegistryDoPublish }}'
+          publishLatest: '${{ env.publishLatest }}'
```
Comment (eriqua) [inline .github/workflows/ms.aad.domainservices.yml:145]: Where is this env variable coming from?

File: .azuredevops/pipelineTemplates/jobs.publishModule.yml [hunk 2]
```diff
@@ -21,12 +21,13 @@
...
 ##   | poolName                        | '$(poolName)'                        | You can provide either a [poolname] or [vmImage] to run the job on                                      | 'Custom Deployment Pool'          |
 ##   | vmImage                         | '$(vmImage)'                         | You can provide either a [poolname] or [vmImage] to run the job on                                      | 'ubuntu20.04'                     |
 ##   | defaultJobTimeoutInMinutes      | 120                                  | The timeout for the job in this pipeline                                                                | 120                               |
 ##   | modulePath                      | '$(modulePath)'                      | The path to the module to deploy. E.g. [c:/KeyVault]                                                    | 'c:/KeyVault'                     |
+##   | publishLatest                   | '$(publishLatest)'                   | Flag to indicate whether or not to publish a "latest" version                                           | true                              |
```
Comment (eriqua) [inline .azuredevops/pipelineTemplates/jobs.publishModule.yml:30]:
```suggestion
##   | publishLatest                   | '$(publishLatest)'                   | Flag to indicate whether or not to publish a "latest" version to Bicep Registry and Template Specs                                           | true                              |
```

File: .github/workflows/ms.aad.domainservices.yml (hunk 1 above)
Comment (AlexanderSehr) [reply .github/workflows/ms.aad.domainservices.yml:145]: The `settings.yml`

File: docs/wiki/Getting started - Scenario 2 Onboard module library and CI environment.md [hunk 3]
```diff
@@ -354,14 +354,16 @@ This section will explain what is required to publish the modules to [Azure Arti
 
 1. An Azure DevOps organization and project
 1. An Azure DevOps artifacts feed
-   > Note: The default feed name is `ResourceModules` as configured in the [`settings.yml`](https://github.com/Azure/ResourceModules/blob/main/settings.yml) file's variable `vstsFeedName`. Update the value here if you want to use a different name, but make sure it matches the name of the artifact feed created in Azure DevOps.
+   > **Note:** The default feed name is `carml` as configured in the [`settings.yml`](https://github.com/Azure/ResourceModules/blob/main/settings.yml) file's variable `vstsFeedName`. Update the value here if you want to use a different name, but make sure it matches the name of the artifact feed created in Azure DevOps.
```
[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply docs/wiki/Getting started - Scenario 2 Onboard module library and CI environment.md:357]: I would however question if the section right below isn't sufficient:
```
#### The dependent components are
(...)
3. An Azure DevOps project to host the artifact feed
   > **Note:** There are a couple options to consider when setting up an Azure Artifact feed. For example, organization-scoped feeds vs project-scoped feeds. Please see what option suits your needs by reviewing the [feeds](https://docs.microsoft.com/en-us/azure/devops/artifacts/concepts/feeds?view=azure-devops) document first.
4. If you chose the feed to be project-scoped, you will need the Project Build Service account to have `Contributor` access to publish to the Azure Artifacts feed. To set this, follow the [Pipeline permission](https://docs.microsoft.com/en-us/azure/devops/artifacts/feeds/feed-permissions?view=azure-devops#pipelines-permissions) steps.
```

[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply docs/wiki/Getting started - Scenario 2 Onboard module library and CI environment.md:357]: Or - if it's not - if we should simply link to the [official guidelines](https://learn.microsoft.com/en-us/azure/devops/artifacts/concepts/feeds?view=azure-devops)

[the remark this answers is not in the document]
Comment (eriqua) [reply docs/wiki/Getting started - Scenario 2 Onboard module library and CI environment.md:357]: For the how to I'd vote for linking to the official guidelines. 
However, when linking to the guidelines, we should explicitly call out that we are by default referencing to a feed named carml, so users should also set that up in that way unless required differently. Then, the note you added here should be added to that paragraph.

[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply docs/wiki/Getting started - Scenario 2 Onboard module library and CI environment.md]: Updated :)

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**